File: colab/benchmark.py
"""
MagneticLM Benchmark - OPTIMIZED with NumPy cache evaluation
"""

import math
import time
import numpy as np
from graph import WordGraph
from trainer import Trainer, tokenize
from typing import List


def compute_perplexity(graph: WordGraph, test_lines: List[str], mode: str = "magnetic") -> float:
    total_log_prob = 0.0
    total_tokens = 0
    floor = 1e-10

    # Cache as list (will be trimmed)
    cache = []

    for idx, line in enumerate(test_lines):
        if (idx + 1) % 500 == 0:
            print(f"\r    [{mode}] {idx+1}/{len(test_lines)}...", end="", flush=True)

        words = tokenize(line)
        if len(words) < 2:
            continue

        is_new = len(cache) < 20

        for i in range(1, len(words)):
            ctx_start = max(0, i - graph.max_order)
            context = tuple(words[ctx_start:i])

            if mode == "bigram":
                key = words[i - 1]
                total = graph.ngram_totals.get(key, 0)
                prob = graph.ngram_counts.get(key, {}).get(words[i], 0) / total if total > 0 else 0
            elif mode == "kn":
                prob = graph.kn_probability(context, words[i])
            else:  # cache or magnetic
                prob = graph.magnetic_probability(context, words[i], cache, is_new)

            prob = max(prob, floor)
            total_log_prob += math.log(prob)
            total_tokens += 1

            cache.append((words[i], context))
            if len(cache) > 4000:
                cache = cache[-4000:]

    print()
    if total_tokens == 0:
        return float('inf')
    return math.exp(-total_log_prob / total_tokens)


def run_benchmark(train_path: str, test_path: str, physics_iterations: int = 50):
    print("=" * 60)
    print("  MagneticLM v6: Modified KN + Physics + Cache + Semantic")
    print("  OPTIMIZED: NumPy vectorized physics")
    print("=" * 60)

    with open(train_path, 'r', encoding='utf-8') as f:
        train_lines = [l.strip() for l in f if l.strip()]
    with open(test_path, 'r', encoding='utf-8') as f:
        test_lines = [l.strip() for l in f if l.strip()]

    print(f"\nTrain: {len(train_lines):,} sentences")
    print(f"Test:  {len(test_lines):,} sentences")

    graph = WordGraph(max_ngram_order=5)
    trainer = Trainer(graph)

    t0 = time.time()
    trainer.train_batch(train_lines)
    graph.build_post_training(physics_iterations)
    train_time = time.time() - t0

    nodes, ngrams, semantic, circles = graph.get_stats()
    print(f"\nTraining: {train_time:.1f}s")
    print(f"Graph: {nodes:,} nodes, {ngrams:,} n-gram, {semantic:,} semantic, {circles:,} circles")
    print(f"Tokens: {graph.total_tokens:,}")

    print("\nComputing perplexity...")

    ppl_bi = compute_perplexity(graph, test_lines, "bigram")
    print(f"  Bigram:                    PPL = {ppl_bi:.1f}")

    ppl_kn = compute_perplexity(graph, test_lines, "kn")
    print(f"  Modified KN-5gram:         PPL = {ppl_kn:.1f}")

    ppl_mag = compute_perplexity(graph, test_lines, "magnetic")
    print(f"  MagneticLM (full):         PPL = {ppl_mag:.1f}")

    print(f"\n{'='*55}")
    print(f"  {'Model':<35}| Perplexity")
    print(f"  {'-'*35}+{'-'*12}")
    print(f"  {'Our Bigram':<35}| {ppl_bi:.1f}")
    print(f"  {'Our Modified KN-5gram':<35}| {ppl_kn:.1f}")
    print(f"  {'Our MagneticLM (full)':<35}| {ppl_mag:.1f}")
    print(f"  {'-'*35}+{'-'*12}")
    print(f"  {'5-gram KN (published)':<35}| ~141")
    print(f"  {'LSTM (Zaremba 2014)':<35}| ~78")
    print(f"  {'AWD-LSTM + Cache (Merity)':<35}| ~52")
    print(f"  {'Transformer-XL (Dai 2019)':<35}| ~54")
    print(f"{'='*55}")

    best = min(ppl_kn, ppl_mag)
    if best < 52: print("\n  >>> BEAT AWD-LSTM + Cache! <<<")
    elif best < 78: print("\n  >>> BEAT LSTM! <<<")
    elif best < 141: print("\n  BETTER than 5-gram KN!")

    return graph, ppl_bi, ppl_kn, ppl_mag


if __name__ == "__main__":
    import sys
    if len(sys.argv) >= 3:
        run_benchmark(sys.argv[1], sys.argv[2],
                      physics_iterations=int(sys.argv[3]) if len(sys.argv) > 3 else 50)
    else:
        print("Usage: python benchmark.py <train.txt> <test.txt> [physics_iterations]")

File: Benchmark.cs
using ACommerce.MagneticLM.Graph;
using ACommerce.MagneticLM.Training;
using System.Diagnostics;

namespace ACommerce.MagneticLM;

public static class Benchmark
{
    public static void Run(string trainPath, string testPath)
    {
        Console.OutputEncoding = System.Text.Encoding.UTF8;
        Console.WriteLine("╔═══════════════════════════════════════════════════╗");
        Console.WriteLine("║  MagneticLM v3: Modified KN + Cache + Semantic   ║");
        Console.WriteLine("╚═══════════════════════════════════════════════════╝\n");

        var trainLines = File.ReadAllLines(trainPath).Where(l => l.Trim().Length > 0).ToArray();
        var testLines = File.ReadAllLines(testPath).Where(l => l.Trim().Length > 0).ToArray();
        Console.WriteLine($"Train: {trainLines.Length:N0} sentences");
        Console.WriteLine($"Test:  {testLines.Length:N0} sentences");

        var graph = new WordGraph { MaxNgramOrder = 5, SemanticThreshold = 0.05, TransitiveDecay = 0.5 };
        var trainer = new Trainer(graph);

        var sw = Stopwatch.StartNew();
        trainer.TrainBatch(trainLines);

        // بناء كل الهياكل بعد التدريب
        graph.BuildPostTrainingStructures();
        sw.Stop();

        var (nodes, ngramEntries, sEdges, circles) = graph.GetStats();
        Console.WriteLine($"\nTraining: {sw.Elapsed.TotalSeconds:F1}s");
        Console.WriteLine($"Graph: {nodes:N0} nodes, {ngramEntries:N0} n-gram, {sEdges:N0} semantic, {circles:N0} circles");
        Console.WriteLine($"Tokens: {graph.TotalTokens:N0}");
        Console.WriteLine($"Discounts: D1={graph.D1:F3} D2={graph.D2:F3} D3+={graph.D3Plus:F3}\n");

        // === Perplexity ===
        Console.Write("Computing...");

        var pplBigram = ComputePerplexity(graph, testLines, mode: "bigram");
        Console.Write($"\r  Bigram:                      PPL = {pplBigram:F1}\n");

        var pplKN = ComputePerplexity(graph, testLines, mode: "kn");
        Console.Write($"  Modified KN-5gram:           PPL = {pplKN:F1}\n");

        var pplCache = ComputePerplexity(graph, testLines, mode: "cache");
        Console.Write($"  Modified KN + Cache:         PPL = {pplCache:F1}\n");

        var pplMagnetic = ComputePerplexity(graph, testLines, mode: "magnetic");
        Console.Write($"  MagneticLM (KN+Cache+Sem):   PPL = {pplMagnetic:F1}\n");

        // === Comparison ===
        Console.WriteLine("\n════════════════════════════════════════════════════");
        Console.WriteLine("  Model                          | Perplexity");
        Console.WriteLine("  ───────────────────────────────┼───────────");
        Console.WriteLine($"  Our Bigram                     | {pplBigram:F1}");
        Console.WriteLine($"  Our Modified KN-5gram          | {pplKN:F1}");
        Console.WriteLine($"  Our KN + Cache                 | {pplCache:F1}");
        Console.WriteLine($"  Our MagneticLM (KN+Cache+Sem)  | {pplMagnetic:F1}");
        Console.WriteLine("  ───────────────────────────────┼───────────");
        Console.WriteLine("  5-gram KN (published)          | ~141");
        Console.WriteLine("  LSTM (Zaremba 2014)            | ~78");
        Console.WriteLine("  AWD-LSTM (Merity 2018)         | ~57");
        Console.WriteLine("  Transformer-XL (Dai 2019)      | ~54");
        Console.WriteLine("════════════════════════════════════════════════════");

        var knImprove = (pplBigram - pplKN) / pplBigram * 100;
        var cacheImprove = (pplKN - pplCache) / pplKN * 100;
        var semImprove = (pplCache - pplMagnetic) / pplCache * 100;
        Console.WriteLine($"\n  KN vs Bigram:      {knImprove:F1}% better");
        Console.WriteLine($"  Cache vs KN:       {cacheImprove:F1}% better");
        Console.WriteLine($"  Semantic vs Cache: {semImprove:F1}% better");
        Console.WriteLine($"  Total vs Bigram:   {(pplBigram - pplMagnetic) / pplBigram * 100:F1}% better");

        if (pplMagnetic < 78) Console.WriteLine("\n  >>> BEAT LSTM! <<<");
        else if (pplMagnetic < 100) Console.WriteLine("\n  Between LSTM and 5-gram KN");
        else if (pplMagnetic < 141) Console.WriteLine("\n  BETTER than 5-gram KN!");
    }

    private static double ComputePerplexity(WordGraph graph, string[] testLines, string mode)
    {
        double totalLogProb = 0;
        int totalTokens = 0;
        double floor = 1e-10;

        // Cache يتراكم عبر الأسطر (مفيد لـ WikiText/مقالات طويلة)
        var cacheEntries = new List<(string Word, string[] Context)>();

        int done = 0;
        foreach (var line in testLines)
        {
            done++;
            if (done % 1000 == 0)
                Console.Write($"\r  [{mode}] {done}/{testLines.Length}...          ");

            var words = Trainer.Tokenize(line);
            if (words.Length < 2) continue;

            bool isNewSentence = (cacheEntries.Count < 20);
            // لا نصفّر الـ cache - يتراكم عبر الجمل (مفيد لـ WikiText)

            for (int i = 1; i < words.Length; i++)
            {
                var currentWord = words[i];
                var contextStart = Math.Max(0, i - graph.MaxNgramOrder);
                var fullContext = words[contextStart..i];

                double prob;
                switch (mode)
                {
                    case "bigram":
                        var biKey = words[i - 1];
                        if (graph.NgramTotals.TryGetValue(biKey, out var biTotal) && biTotal > 0)
                        {
                            graph.NgramCounts[biKey].TryGetValue(currentWord, out var biCount);
                            prob = biCount / biTotal;
                        }
                        else prob = 0;
                        break;

                    case "kn":
                        prob = graph.GetInterpolatedProbability(fullContext, currentWord);
                        break;

                    case "cache":
                        prob = graph.GetMagneticProbability(fullContext, currentWord,
                            cacheEntries, isNewSentence: isNewSentence);
                        break;

                    case "magnetic":
                        prob = graph.GetMagneticProbability(fullContext, currentWord,
                            cacheEntries, isNewSentence: isNewSentence);
                        break;

                    default:
                        prob = 0;
                        break;
                }

                prob = Math.Max(prob, floor);
                totalLogProb += Math.Log(prob);
                totalTokens++;
                cacheEntries.Add((currentWord, fullContext));
                // حد أقصى للـ cache (يحتفظ بآخر 500 فقط)
                if (cacheEntries.Count > 4000)
                    cacheEntries.RemoveRange(0, cacheEntries.Count - 4000);
            }
        }

        return totalTokens > 0 ? Math.Exp(-totalLogProb / totalTokens) : double.MaxValue;
    }
}
